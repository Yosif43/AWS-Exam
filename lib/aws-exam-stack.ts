import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Stack, StackProps, Duration, RemovalPolicy } from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as iam from 'aws-cdk-lib/aws-iam';



export class AwsExamStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.Table(this, 'InvalidJsonTable', {
      partitionKey: {
        name: 'PK',
        type: dynamodb.AttributeType.STRING
      },
      sortKey: {
        name: 'SK',
        type: dynamodb.AttributeType.STRING
      },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      timeToLiveAttribute: 'ttl',
      removalPolicy: RemovalPolicy.DESTROY,
      stream: dynamodb.StreamViewType.NEW_AND_OLD_IMAGES,
    });

    const notifyEmail = 'yosifyosifov4393@gmail.com';

    const handler = new lambda.Function(this, 'JsonValidatorFunction', {
      runtime: lambda.Runtime.NODEJS_18_X,
      code: lambda.Code.fromAsset('src'),
      handler: 'src/handler.ts',
      environment: {
        TABLE_NAME: table.tableName,
        NOTIFY_EMAIL: notifyEmail,
        TTL_MINUTES: '1440',
      },
    });

    const onDeleteHandler = new lambda.Function(this, 'DynamoDeleteHandler', {
      runtime: lambda.Runtime.NODEJS_18_X,
      code: lambda.Code.fromAsset('src'),
      handler: 'src/onDeleteHandler',
      environment: {
        NOTIFY_EMAIL: notifyEmail,
      },
    });

    table.grantReadWriteData(handler);
    table.grantStreamRead(onDeleteHandler);

    handler.addToRolePolicy(new iam.PolicyStatement({
      actions: ['ses:SendEmail', 'ses:SendRawEmail'],
      resources: ['*'],
    }));

    onDeleteHandler.addToRolePolicy(new iam.PolicyStatement({
      actions: ['ses:SendEmail', 'ses:SendRawEmail'],
      resources: ['*'],
    }));

    onDeleteHandler.addEventSourceMapping('StreamTrigger', {
      eventSourceArn: table.tableStreamArn,
      startingPosition: lambda.StartingPosition.TRIM_HORIZON,
    });

    new apigateway.LambdaRestApi(this, 'Endpoint', {
      handler,
      deployOptions: {
        stageName: 'prod',
      },
    });

  }
}
