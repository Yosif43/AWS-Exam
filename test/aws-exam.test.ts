import {Template} from "aws-cdk-lib/assertions";
import {App} from "aws-cdk-lib";
import {AwsExamStack} from "../lib/aws-exam-stack";

test('SQS Queue Created', () => {
   const app = new App();
   const stack = new AwsExamStack(app, 'MyTestStack');
   const template = Template.fromStack(stack);

   expect(template).toMatchSnapshot();

});
