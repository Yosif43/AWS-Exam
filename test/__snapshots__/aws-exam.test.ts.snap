// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS Queue Created 1`] = `
{
  "Outputs": {
    "Endpoint8024A810": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "EndpointEEF1FD8F",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "EndpointDeploymentStageprodB78BEEA0",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DynamoDeleteHandlerDB39441B": {
      "DependsOn": [
        "DynamoDeleteHandlerServiceRoleDefaultPolicyBDC30D7E",
        "DynamoDeleteHandlerServiceRoleC918C743",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "98929edc65342f46016e83fe8cb51b8c66f41f09bbc0401abc62c48dfe461c30.zip",
        },
        "Environment": {
          "Variables": {
            "NOTIFY_EMAIL": "yosifyosifov4393@gmail.com",
          },
        },
        "Handler": "src/onDeleteHandler",
        "Role": {
          "Fn::GetAtt": [
            "DynamoDeleteHandlerServiceRoleC918C743",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DynamoDeleteHandlerServiceRoleC918C743": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DynamoDeleteHandlerServiceRoleDefaultPolicyBDC30D7E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "InvalidJsonTable734F1BFF",
                  "StreamArn",
                ],
              },
            },
            {
              "Action": [
                "ses:SendEmail",
                "ses:SendRawEmail",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DynamoDeleteHandlerServiceRoleDefaultPolicyBDC30D7E",
        "Roles": [
          {
            "Ref": "DynamoDeleteHandlerServiceRoleC918C743",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DynamoDeleteHandlerStreamTriggerA490A978": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "InvalidJsonTable734F1BFF",
            "StreamArn",
          ],
        },
        "FunctionName": {
          "Ref": "DynamoDeleteHandlerDB39441B",
        },
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "EndpointANY485C938B": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "JsonValidatorFunction6EEA5C5F",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "EndpointEEF1FD8F",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "EndpointEEF1FD8F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "EndpointANYApiPermissionMyTestStackEndpoint695FDFAAANY7AA09D8D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "JsonValidatorFunction6EEA5C5F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "EndpointEEF1FD8F",
              },
              "/",
              {
                "Ref": "EndpointDeploymentStageprodB78BEEA0",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "EndpointANYApiPermissionTestMyTestStackEndpoint695FDFAAANY75707C47": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "JsonValidatorFunction6EEA5C5F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "EndpointEEF1FD8F",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "EndpointAccountB8304247": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "EndpointEEF1FD8F",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "EndpointCloudWatchRoleC3C64E0F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "EndpointCloudWatchRoleC3C64E0F": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "EndpointDeployment318525DA7f2d74829180926330650b1345f9501b": {
      "DependsOn": [
        "EndpointproxyANYC09721C5",
        "Endpointproxy39E2174E",
        "EndpointANY485C938B",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "EndpointEEF1FD8F",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "EndpointDeploymentStageprodB78BEEA0": {
      "DependsOn": [
        "EndpointAccountB8304247",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "EndpointDeployment318525DA7f2d74829180926330650b1345f9501b",
        },
        "RestApiId": {
          "Ref": "EndpointEEF1FD8F",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "EndpointEEF1FD8F": {
      "Properties": {
        "Name": "Endpoint",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "Endpointproxy39E2174E": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "EndpointEEF1FD8F",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "EndpointEEF1FD8F",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "EndpointproxyANYApiPermissionMyTestStackEndpoint695FDFAAANYproxyA3DADBF6": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "JsonValidatorFunction6EEA5C5F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "EndpointEEF1FD8F",
              },
              "/",
              {
                "Ref": "EndpointDeploymentStageprodB78BEEA0",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "EndpointproxyANYApiPermissionTestMyTestStackEndpoint695FDFAAANYproxy90B88BB5": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "JsonValidatorFunction6EEA5C5F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "EndpointEEF1FD8F",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "EndpointproxyANYC09721C5": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "JsonValidatorFunction6EEA5C5F",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "Endpointproxy39E2174E",
        },
        "RestApiId": {
          "Ref": "EndpointEEF1FD8F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "InvalidJsonTable734F1BFF": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TimeToLiveSpecification": {
          "AttributeName": "ttl",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "JsonValidatorFunction6EEA5C5F": {
      "DependsOn": [
        "JsonValidatorFunctionServiceRoleDefaultPolicy116F24E0",
        "JsonValidatorFunctionServiceRoleC5009459",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "98929edc65342f46016e83fe8cb51b8c66f41f09bbc0401abc62c48dfe461c30.zip",
        },
        "Environment": {
          "Variables": {
            "NOTIFY_EMAIL": "yosifyosifov4393@gmail.com",
            "TABLE_NAME": {
              "Ref": "InvalidJsonTable734F1BFF",
            },
            "TTL_MINUTES": "1440",
          },
        },
        "Handler": "src/handler.ts",
        "Role": {
          "Fn::GetAtt": [
            "JsonValidatorFunctionServiceRoleC5009459",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "JsonValidatorFunctionServiceRoleC5009459": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "JsonValidatorFunctionServiceRoleDefaultPolicy116F24E0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "InvalidJsonTable734F1BFF",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": [
                "ses:SendEmail",
                "ses:SendRawEmail",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JsonValidatorFunctionServiceRoleDefaultPolicy116F24E0",
        "Roles": [
          {
            "Ref": "JsonValidatorFunctionServiceRoleC5009459",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
